src/ApiService/ApiService/onefuzzlib/VmssOperations.cs

```diff
@@ -444,22 +428,43 @@ public Async.Task<IReadOnlyList<string>> ListAvailableSkus(Region region)
             return skuNames;
         });
 
-    public async Async.Task<OneFuzzResultVoid> ReimageNodes(Guid scalesetId, IReadOnlySet<Guid> machineIds) {
-        var result = await CheckCanUpdate(scalesetId);
-        if (!result.IsOk) {
-            return OneFuzzResultVoid.Error(result.ErrorV);
-        }
+    private async Async.Task<HashSet<string>> ResolveInstanceIds(Guid scalesetId, IEnumerable<Node> nodes) {
+
+        // only initialize this if we find a missing InstanceId
+        var machineToInstanceLazy = new Lazy<Task<IDictionary<Guid, string>>>(async () => {
+            var machineToInstance = await ListInstanceIds(scalesetId);
+            if (machineToInstance is null) {
+                throw new Exception($"cannot find nodes in scaleset {scalesetId}: scaleset does not exist");
+            }
+
+            return machineToInstance;
+        });
 
         var instanceIds = new HashSet<string>();
-        var machineToInstance = await ListInstanceIds(scalesetId);
-        foreach (var machineId in machineIds) {
-            if (machineToInstance.TryGetValue(machineId, out var instanceId)) {
-                _ = instanceIds.Add(instanceId);
+        foreach (var node in nodes) {
+            if (node.InstanceId is not null) {
+                _ = instanceIds.Add(node.InstanceId);
+                continue;
+            }
+
+            var lookup = await machineToInstanceLazy.Value;
```

Comment:
suggestion:
we could use this logic to update the instanceId in the node entry in the table. 
I would do it in separate method but that is not a strong preference.

Reply:
This is just here for back-compat. After the next release we should never see missing InstanceID here.

Reply:
I've made a change in the commit e41a2e9e47cdc6c50b761ea8b72d41ea6e752d2f to insert InstanceID during the CleanupNodes phase, so now there's never a case where we insert a Node without an InstanceID.